src/Native/Runtime/arm/WriteBarriers.S

```diff
@@ -2,67 +2,375 @@
+          cmp          r12, \REFREG
+          bne          LOCAL_LABEL(\BASENAME\()_UpdateShadowHeap_Invalidate_\REFREG)
+
+          // The original DESTREG value is now restored but the stack has a value (the shadow version of the
+          // location) pushed. Need to discard this push before we are done.
+          add          sp, #4
+          b            LOCAL_LABEL(\BASENAME\()_UpdateShadowHeap_Done_\REFREG)
+
+LOCAL_LABEL(\BASENAME\()_UpdateShadowHeap_Invalidate_\REFREG):
+          // Someone went and updated the real heap. We need to invalidate the shadow location since we can't
+          // guarantee whose shadow update won.
+
+          // Retrieve shadow location from the stack and restore original DESTREG to the stack. This is an
+          // additional memory barrier we don't require but it's on the rare path and x86 doesn't have an xchg
+          // variant that doesn't implicitly specify the lock prefix. Note that INVALIDGCVALUE is a 32-bit
+          // immediate and therefore must be moved into a register before it can be written to the shadow
+          // location.
+          mov          r12, \DESTREG
+          ldr          \DESTREG, [sp]
+          str          r12, [sp]
+          push         \REFREG
+          movw         \REFREG, #(INVALIDGCVALUE & 0xFFFF)
+          movt         \REFREG, #(INVALIDGCVALUE >> 16)
+          str          \REFREG, [\DESTREG]
+          pop          \REFREG
+
+LOCAL_LABEL(\BASENAME\()_UpdateShadowHeap_PopThenDone_\REFREG):
+          // Restore original DESTREG value from the stack.
+          pop          \DESTREG
+
+LOCAL_LABEL(\BASENAME\()_UpdateShadowHeap_Done_\REFREG):
+
+.endm
+
+#else  // WRITE_BARRIER_CHECK
+
+.macro UPDATE_GC_SHADOW BASENAME, REFREG, DESTREG
+.endm
+
+#endif // WRITE_BARRIER_CHECK
+
+// There are several different helpers used depending on which register holds the object reference. Since all
+// the helpers have identical structure we use a macro to define this structure. Two arguments are taken, the
+// name of the register that points to the location to be updated and the name of the register that holds the
+// object reference (this should be in upper case as it's used in the definition of the name of the helper).
+.macro DEFINE_UNCHECKED_WRITE_BARRIER_CORE BASENAME, REFREG
+
+          // Update the shadow copy of the heap with the same value just written to the same heap. (A no-op unless
+          // we're in a debug build and write barrier checking has been enabled).
+          UPDATE_GC_SHADOW \BASENAME, \REFREG, r0
+
+          // If the reference is to an object that's not in an ephemeral generation we have no need to track it
+          // (since the object won't be collected or moved by an ephemeral collection).
+          ldr	         r12, =C_FUNC(g_ephemeral_low)
+          ldr          r12, [r12]
+          cmp          \REFREG, r12
+          blo          LOCAL_LABEL(\BASENAME\()_NoBarrierRequired_\REFREG)
```

Comment:
The target of this conditional branch is unconditional branch. It can be short-circuited to branch directly to `LOCAL_LABEL(\BASENAME\()_EXIT_\REFREG)`